src/net/client/cache.rs

```diff
@@ -0,0 +1,1177 @@
+    async fn cache_lookup(
+        &self,
+        key: &Key,
+    ) -> Result<Option<Arc<Value<C>>>, Error> {
+        // There are 4 flags that may affect the response to a query.
+        // In some cases the response to one value of a flag could be
+        // used for the other value.
+        // This function takes all 4 flags. First we take care of the CD flag.
+        // This flag has to be used as is, so there is not much to do. Next
+        // we pass the request to a function that looks at RD, DO, and AD.
+        self.cache_lookup_rd_do_ad(key).await
+    }
+
+    /// Try to find an cache entry for the key taking into account the
+    /// RD, DO, and AD flags. The CD flag is kept unchanged.
+    async fn cache_lookup_rd_do_ad(
+        &self,
+        key: &Key,
+    ) -> Result<Option<Arc<Value<C>>>, Error> {
+        // For RD=1 we can only use responses to queries with RD set.
+        // For RD=0, first try with RD=0 and then try with RD=1. If
+        // RD=1 has an answer, store it as an answer for RD=0.
+        let opt_value = self.cache_lookup_do_ad(key).await?;
+        if opt_value.is_some() || key.rd {
+            return Ok(opt_value);
+        }
+
+        // Look if there is something with RD=1. We can use the
+        // response unmodified.
+        let mut alt_key = key.clone();
+        alt_key.rd = true;
+        let opt_value = self.cache_lookup_do_ad(&alt_key).await?;
+        if let Some(value) = &opt_value {
+            match &value.response {
+                Err(_) => {
+                    // Just insert the error
+                    self.cache_insert(key.clone(), value.clone()).await;
+                }
+                Ok(msg) => {
+                    let msg = clear_rd(msg)?;
+                    let value =
+                        Arc::new(Value::<C>::new_from_value_and_response(
+                            value.clone(),
+                            Ok(msg),
+                            &self.config,
+                        )?);
+                    self.cache_insert(key.clone(), value.clone()).await;
+                    return Ok(Some(value));
+                }
+            }
+        }
+        Ok(opt_value)
+    }
+
+    /// Try to find an cache entry for the key taking into account the
+    /// DO and AD flags. The CD and RD flags are kept unchanged.
+    async fn cache_lookup_do_ad(
+        &self,
+        key: &Key,
+    ) -> Result<Option<Arc<Value<C>>>, Error> {
+        // For DO=1 we can only use responses to queries with DO set.
+        // For DO=0, first try with DO=0 and then try with DO=1. If
+        // DO=1 has an answer, remove DNSSEC related resource records.
+        // If AD is clear then clear the AD bit.
+
+        // If DO is set then AD is irrelevant. Force AD to be set for
+        // consistency (if DO is set then with respect to the AD flag
+        // the behavior is as if AD is set).
+
+        if key.dnssec_ok {
+            assert!(key.ad);
```

Comment:
This feels a bit scary. Maybe have an enum that reflects the valid combinations of DO and AD bits? Or just force the AD bit to false here?